# Phân loại Bayes Classification Scikit-Learn

In [1]:
import matplotlib.pyplot as plt
import pandas as pd
from sklearn import datasets
from sklearn.model_selection import train_test_split
from sklearn.naive_bayes import GaussianNB

Bộ dữ liệu Iris

Trong bài tập này, chúng ta phải phân loại ba loại hoa diên vĩ dựa trên chiều dài và chiều rộng của lá đài và
cánh hoa. Các loại này bao gồm Setosa, Versicolor và Virginica, được mã hóa lần lượt bằng 0, 1 và 2.

Bộ dữ liệu phân loại diên vĩ có sẵn trong gói scikit-learn. Để tải xuống và sử dụng cho thí nghiệm của chúng ta, bạn có thể triển khai nó bằng mã sau:

In [2]:
# Downloading the iris dataset
iris = datasets.load_iris()
class_names = iris.target_names
# Convert to pandas format for easier processing
iris_df = pd.DataFrame(iris.data, columns=iris.feature_names)
iris_df['target'] = iris.target

In [3]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)  \
0                5.1               3.5                1.4               0.2   
1                4.9               3.0                1.4               0.2   
2                4.7               3.2                1.3               0.2   
3                4.6               3.1                1.5               0.2   
4                5.0               3.6                1.4               0.2   

   target  
0       0  
1       0  
2       0  
3       0  
4       0

Đầu tiên, chúng tôi chia ngẫu nhiên tập dữ liệu thành tập huấn luyện (80%) và tập thử nghiệm (20%).

In [4]:
# Splitting dataset into train set and test set
X_train, X_test, y_train, y_test = train_test_split(
iris_df[['sepal length (cm)', 'sepal width (cm)',
'petal length (cm)','petal width (cm)']],
iris_df['target'],
test_size=0.2,
random_state=2021)

Mô hình hóa và Kết quả

Trong thí nghiệm này, chúng tôi xây dựng một Bộ phân loại Naive Bayes để phân loại ba loại mống mắt. Chúng tôi giả định chiều dài của
lá đài và cánh hoa tuân theo phân phối chuẩn. Vì vậy, chúng tôi có thể huấn luyện một Bộ phân loại Naive Bayes theo chuẩn Gauss như sau:

In [6]:
# Import Naive Bayes
# In this case, I assume features follow a normal distribution
NB = GaussianNB()
# Train Naive Bayes Model
NB.fit(X_train, y_train)
print("Training accuracy: {:.2f}".format(NB.score(X_train, y_train)))
# Evaluate Model on test set
y_predict = NB.predict(X_test)
print("Testing accuracy NB: {:.2f}".format(NB.score(X_test, y_test)))

Training accuracy: 0.97
Testing accuracy NB: 0.93


Sau khi đào tạo, mô hình của chúng tôi đạt độ chính xác 93% trên tập thử nghiệm.

Hình ảnh hóa kết quả

Mô hình của chúng tôi rất hiệu quả trong việc phân loại các loài hoa diên vĩ. Để kiểm chứng dự đoán, chúng tôi sử dụng Phân tích Thành phần Chính (PCA) để thu nhỏ các chiều thành mô hình 2 chiều phục vụ cho việc hình ảnh hóa.